File: T-Bank/Algorithms-and-Data-Structures-2025/1sprint-IntroductionToSorting-BinarySearch/contest/H-Зеркальный_код/solution.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>


void solution() {
    /*
    Time Complexity O(N)
    Memory Complexity O(N)
    */
    int n;
    std::cin >> n;
    std::string letters;
    std::cin >> letters;
    std::vector<int> chunks(26, 0);
    for (char& letter : letters) {
        ++chunks[letter - 'A'];
    }
    std::string output = "";
    char median = '0';
    for (int i = 0; i < 26; ++i) {
        if (!chunks[i]) continue;
        if ((chunks[i] % 2 != 0) && (median == '0')) median = (char)(i + 65);
        if (chunks[i] / 2)
            for (int j = 0; j < chunks[i] / 2; j++)
                output += (char)(i + 65);
    }
    std::string palindrom = output;
    if (median != '0') palindrom.push_back(median);
    std::reverse(output.begin(), output.end());
    palindrom += output;
    std::cout << palindrom << "\n";
}


int main() {
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/1sprint-IntroductionToSorting-BinarySearch/contest/H-Зеркальный_код/solution.py
import sys


def solution() -> int:
    """
    Time Complexity O(N)
    Memory Complexity O(N)
    """
    n: int = int(sys.stdin.readline().rstrip())
    letters: str = sys.stdin.readline().rstrip()
    chunks: list[int] = [0] * 26
    for letter in letters:
        chunks[ord(letter) - 65] += 1
    output: list[str] = []
    median: str = ''
    for i in range(26):
        if chunks[i]:
            if chunks[i] & 1 and not median: median = chr(i + 65)
            output.append(chr(i + 65) * (chunks[i] // 2))
    palindrom: str = ''.join(output)
    if median: palindrom += median
    palindrom += ''.join(reversed(output))
    sys.stdout.write(f'{palindrom}\n')


if __name__ == '__main__':
    solution()